### PR TITLE
Feature/at least support search

### PR DESCRIPTION
**Pull request by Dronplane**
### Scope & Purpose

Adds support of AT LEAST quantifier to SEARCH/inverted index.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR: https://github.com/arangodb/enterprise/pull/990
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Review by gnusi — APPROVED** [review]

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 1]
```diff
@@ -1044,10 +1044,53 @@ Result fromRange(irs::boolean_filter* filter, QueryContext const& /*ctx*/,
...
 std::pair<Result, aql::AstNodeType> buildBinaryArrayComparisonPreFilter(
     irs::boolean_filter*& filter, aql::AstNodeType arrayComparison,
-    const aql::AstNode* quantifierNode, size_t arraySize) {
+    const aql::AstNode* quantifierNode, size_t arraySize,
+    QueryContext const& ctx) {
   TRI_ASSERT(quantifierNode);
   auto quantifierType =
       static_cast<aql::Quantifier::Type>(quantifierNode->getIntValue(true));
+
+  int64_t atLeastCount{1};
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1053]: Why is 1 the default?

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 2]
```diff
@@ -1044,10 +1044,53 @@ Result fromRange(irs::boolean_filter* filter, QueryContext const& /*ctx*/,
 
 std::pair<Result, aql::AstNodeType> buildBinaryArrayComparisonPreFilter(
     irs::boolean_filter*& filter, aql::AstNodeType arrayComparison,
-    const aql::AstNode* quantifierNode, size_t arraySize) {
+    const aql::AstNode* quantifierNode, size_t arraySize,
+    QueryContext const& ctx) {
   TRI_ASSERT(quantifierNode);
   auto quantifierType =
       static_cast<aql::Quantifier::Type>(quantifierNode->getIntValue(true));
+
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1052]: `aql::` not necessary here

File: arangod/IResearch/IResearchFilterFactory.cpp (hunk 1 above)
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1053]: NIT
what about make two vars size_t atLeastCount and int64_t result? in that case most static_cast will be not necessary

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 3]
```diff
@@ -1185,45 +1234,64 @@ std::pair<Result, aql::AstNodeType> buildBinaryArrayComparisonPreFilter(
...
+            if (filter) {
+              filter = static_cast<irs::boolean_filter*>(
+                  &filter->add<irs::Or>().min_match_count(
+                      static_cast<size_t>(atLeastCount)));
+            }
+            expansionNodeType = aql::NODE_TYPE_OPERATOR_BINARY_NE;
+            break;
+          default:
+            TRI_ASSERT(false);  // new array comparison operator?
+            return std::make_pair(
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1289]:
```suggestion
            return std::pair{
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 4]
```diff
@@ -1241,10 +1309,16 @@ class ByTermSubFilterFactory {
...
                                 FilterContext const& filterCtx) {
-    TRI_ASSERT(aql::NODE_TYPE_OPERATOR_BINARY_EQ == node.cmp);
+    TRI_ASSERT(aql::NODE_TYPE_OPERATOR_BINARY_EQ == node.cmp ||
+               aql::NODE_TYPE_OPERATOR_BINARY_NE == node.cmp);
     irs::by_term* termFilter = nullptr;
     if (filter) {
-      termFilter = &filter->add<irs::by_term>();
+      if (aql::NODE_TYPE_OPERATOR_BINARY_NE == node.cmp) {
+        termFilter =
+            &filter->add<irs::And>().add<irs::Not>().filter<irs::by_term>();
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1318]: Can you add comment about why we need this workaround

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 5]
```diff
@@ -1255,10 +1329,16 @@ class ByTermSubFilterFactory {
...
                                  QueryContext const& ctx,
                                  FilterContext const& filterCtx) {
-    TRI_ASSERT(aql::NODE_TYPE_OPERATOR_BINARY_EQ == arrayExpansionNodeType);
+    TRI_ASSERT(aql::NODE_TYPE_OPERATOR_BINARY_EQ == arrayExpansionNodeType ||
+               aql::NODE_TYPE_OPERATOR_BINARY_NE == arrayExpansionNodeType);
     irs::by_term* termFilter = nullptr;
     if (filter) {
-      termFilter = &filter->add<irs::by_term>();
+      if (aql::NODE_TYPE_OPERATOR_BINARY_NE == arrayExpansionNodeType) {
+        termFilter =
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1337]: Can you add comment about why we need this workaround

**Review by MBkkt — APPROVED** [review]

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 6]
```diff
@@ -1044,10 +1044,53 @@ Result fromRange(irs::boolean_filter* filter, QueryContext const& /*ctx*/,
 
 std::pair<Result, aql::AstNodeType> buildBinaryArrayComparisonPreFilter(
```
[the remark this answers is not in the document]
Comment (Dronplane) [reply arangod/IResearch/IResearchFilterFactory.cpp:1045]: Good point. I guess this code was written before ResultT was introduced. We should consider rewriting this in the follow up PRs

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 7]
```diff
@@ -1044,10 +1044,53 @@ Result fromRange(irs::boolean_filter* filter, QueryContext const& /*ctx*/,
...
+        filter->add<irs::empty>();
+      }
+      return {Result(), aql::NODE_TYPE_ROOT};
+    } else if (static_cast<size_t>(atLeastCount) == arraySize) {
+      quantifierType = aql::Quantifier::Type::kAll;
+    } else if (arrayComparison != aql::NODE_TYPE_OPERATOR_BINARY_ARRAY_NIN &&
+               arrayComparison != aql::NODE_TYPE_OPERATOR_BINARY_ARRAY_NE) {
+      // processing  of AT LEAST is different only for NIN/NE case
+      quantifierType = aql::Quantifier::Type::kAny;
+    }
```
[the remark this answers is not in the document]
Comment (Dronplane) [reply arangod/IResearch/IResearchFilterFactory.cpp:1091]: There is no "else". If conditions are not fulfilled atLeast is processed as is

File: arangod/IResearch/IResearchFilterFactory.cpp (hunk 1 above)
Comment (Dronplane) [reply arangod/IResearch/IResearchFilterFactory.cpp:1053]: To use with  kAny. As kAny is   atLeast(1)

**Comment by Dronplane** [comment]
Tests blue except unrelated failures on windows

**Pull request merged**